lib/std/posix.zig

```diff
@@ -4344,6 +4386,51 @@ pub fn fstatat(dirfd: fd_t, pathname: []const u8, flags: u32) FStatAtError!Stat
     }
 }
 
+// MKDEV in the kernel.
+inline fn makedevLinux(major: u32, minor: u32) dev_t {
+    return (major << 20) | minor;
+}
+
+// Perform fstatat() by way of statx() for systems that only have the latter (e.g. linux-riscv32).
+// More or less mirrors what glibc and musl do for platforms like this.
+fn fstatatStatxZ(dirfd: fd_t, pathname: [*:0]const u8, stat_buf: *Stat, flags: u32) usize {
+    var statx = mem.zeroes(linux.Statx);
+    const rc = linux.statx(dirfd, pathname, flags | AT.NO_AUTOMOUNT, linux.STATX_BASIC_STATS, &statx);
+    const rcs: isize = @bitCast(rc);
+
+    if (rcs > -4096) return rc;
+
+    stat_buf.* = .{
+        // Convert device values to the older format.
+        .dev = makedevLinux(statx.dev_major, statx.dev_minor),
+        .rdev = makedevLinux(statx.rdev_major, statx.rdev_minor),
```

Comment:
Instead of emulating an older POSIX API in this layer, the fstatat function should compile error on riscv since it does not actually exist. The fstatx function in the posix layer just does the normal wrapping for fstatx. And then in the `std.fs` cross-platform abstraction layer, it can convert directly from `statx` to `std.fs` rather than taking a detour through the nonexistent fstatat API.

Reply:
> The fstatx function in the posix layer just does the normal wrapping for fstatx.

Well, that's the thing - `statx` is *not* POSIX, and we do not have a wrapper for it in `std.posix`. That's not necessarily the end of the world, but it does raise the question of what to actually do in `std.fs`.

Reply:
In fact, the problem goes deeper; `std.fs` exposes `Stat` in a few places (e.g. `std.fs.File.Stat.fromSystem()`. That's why I went the way I did here.

Reply:
The key design criteria is to don't create a fake `Stat` struct. We have `std.fs.File.Stat` to abstract over the real OS level stat.

You can put the statx wrapper in std.os instead if you want.

Perhaps instead of `std.fs.File.Stat.fromSystem` a better function would be `Statx.toStd` and `Statx.fromStd`, and equivalent functions on `Stat`.

Reply:
`std.fs.File.MetadataUnix` notably also has an `std.posix.Stat`-typed field, and it seems like that might be deliberate considering `std.fs.File.Stat` exists? Judging by this and `MetadataWindows`, the intention seems to be to expose all the native details as-is.

Ok, taking a step back for a moment: If Zig can take an official stance on #20423, we could simply make `std.fs` always use `Statx`/`statx()` on Linux (it's a better API anyway), and `std.posix.Stat`/`std.posix.fstatat()` on other POSIX systems. Alternatively, if we really want to support older Linux versions than `std.Target` claims we do, `std.fs` could use `statx` only if `@TypeOf(std.posix.Stat) == void`. Do either of these sound reasonable?

Reply:
Holy crap I'm just now seeing https://github.com/ziglang/zig/pull/10486 for the first time. I will go do a retroactive review on that code.

Always using statx sounds like the right move to me.

Reply:
Wait, I just noticed that `std.fs` actually *does* use `statx` for `MetadataLinux`! But, it also contains code that *still* uses `fstatat` on Linux for `std.fs.File.Stat` specifically. Strange situation. I'll clean it up.